rfcs/2021-06-02-7709-helm-deprecate-custom-config-keys.md

```diff
@@ -0,0 +1,90 @@
+# RFC 7709 - 2021-06-02 - Helm: Deprecate "custom" configuration keys
+
+Today the templates that user's leverage to generate their Vector config are separate from the ones used to generate the "default" Vector configuration.
+While these are not very complex in their functionality they do duplicate much of the same logic.
+
+## Scope
+
+This RFC will cover only the existing "default" `sinks` and `sources` defined in our Helm charts today.
+
+vector-agent:
+
+- `kubernetes_logs`
+- `vector`
+- `internal_metrics`
+- `host_metrics`
+- `prometheus_exporter`
+
+vector-aggregator:
+
+- `vector`
+- `internal_metrics`
+- `prometheus_exporter`
+
+## Motivation
+
+We're duplicating the templating used to generate Vector's configuration file in Helm based deployments. While I was working on transitioning the
+currently TOML config to YAML realized the configuration building had several unexpected complications due to the separation between Vector's provided
+"default" configurations (listed in the Scope section) and a user provided configuration located under the `sources`, `transforms`, and `sinks` keys.
+
+## Internal Proposal
+
+The "custom" `values.yaml` keys be deprecated and replaced with default values under the `sources`, `transforms`, and `sinks` keys that users configure
+for their provided Vector configuration. An example is below:
+
+```yaml
+-  internalMetricsSource:
+-    enabled: true
+-    sourceId: internal_metrics
+-    config: {}
+-  hostMetricsSource:
+-    enabled: true
+-    sourceId: host_metrics
+-    config:
+-      filesystem:
+-        devices:
+-          excludes: [binfmt_misc]
+-        filesystems:
+-          excludes: [binfmt_misc]
+-        mountpoints:
+-          excludes: ["*/proc/sys/fs/binfmt_misc"]
++  sources:
++    internal_metrics:
++      type: internal_metrics
++    host_metrics:
++      type: host_metrics
++      filesystem:
++        devices:
++          excludes: ["binfmt_misc"]
++        filesystems:
++          excludes: ["binfmt_misc"]
++        mountpoints:
++          excludes: ["*/proc/sys/fs/binfmt_misc"]
+```
+
+## Doc-level Proposal
+
+The only page we currently have with documentation around Helm is on the Kubernetes platform installation page, which today does not mention how to configure
+any of the default components.
+
+## Rationale
+
+This has us use the same configuration levers as our users do, and unifies _how_ Vector can be configured in Kubernetes. The default values can also be used
+as "in-line" documentation users can refer to as an example configuration. This also reduces the amount of changes needed if we need to adjust configuration
+templating in the future, reducing some long term burden (slight, and I can't imagine that will change much, or ever, once we stop converting the config into TOML).
+
+## Drawbacks
+
+This change does make opting in or out to our default configurations slightly more difficult, since there would no longer be an "enabled" toggle. To opt out users
```

Comment:
What would disabling look after this change? I note from the [setup instructions](https://vector.dev/docs/setup/installation/platforms/kubernetes/#add-the-vector-repo) it seems like users just provide additional YAML that seems to be merged into the `values.yaml` that we ship. I'm not seeing how they would delete configuration in this way.

Reply:
🤔 I would have expected passing in a custom values file with a `source:` key would overwrite the default values we provide - I wonder if that has something to do with how we're templating these today. I'll do some research

Reply:
Ugh, you can't overwrite the whole object under `sources` - you can overwrite the nested fields though. I'll need to think if there's a way to replace the entire object :/